components/uniset/src/provider.rs

```diff
@@ -14,74 +14,891 @@ pub mod key {
+    pub const GRAPHEME_CLUSTER_BREAK_LF_V1: ResourceKey = resource_key!(uniset, "GCB=LF", 1);
+    pub const GRAPHEME_CLUSTER_BREAK_LV_V1: ResourceKey = resource_key!(uniset, "GCB=LV", 1);
+    pub const GRAPHEME_CLUSTER_BREAK_LVT_V1: ResourceKey = resource_key!(uniset, "GCB=LVT", 1);
+    pub const GRAPHEME_CLUSTER_BREAK_PREPEND_V1: ResourceKey = resource_key!(uniset, "GCB=PP", 1);
+    pub const GRAPHEME_CLUSTER_BREAK_REGIONAL_INDICATOR_V1: ResourceKey =
+        resource_key!(uniset, "GCB=RI", 1);
+    pub const GRAPHEME_CLUSTER_BREAK_SPACINGMARK_V1: ResourceKey =
+        resource_key!(uniset, "GCB=SM", 1);
+    pub const GRAPHEME_CLUSTER_BREAK_T_V1: ResourceKey = resource_key!(uniset, "GCB=T", 1);
+    pub const GRAPHEME_CLUSTER_BREAK_V_V1: ResourceKey = resource_key!(uniset, "GCB=V", 1);
+    pub const GRAPHEME_CLUSTER_BREAK_OTHER_V1: ResourceKey = resource_key!(uniset, "GCB=XX", 1);
+    pub const GRAPHEME_CLUSTER_BREAK_ZWJ_V1: ResourceKey = resource_key!(uniset, "GCB=ZWJ", 1);
+    pub const HANGUL_SYLLABLE_TYPE_LEADING_JAMO_V1: ResourceKey = resource_key!(uniset, "hst=L", 1);
+    pub const HANGUL_SYLLABLE_TYPE_LV_SYLLABLE_V1: ResourceKey = resource_key!(uniset, "hst=LV", 1);
+    pub const HANGUL_SYLLABLE_TYPE_LVT_SYLLABLE_V1: ResourceKey =
+        resource_key!(uniset, "hst=LVT", 1);
+    pub const HANGUL_SYLLABLE_TYPE_NOT_APPLICABLE_V1: ResourceKey =
+        resource_key!(uniset, "hst=NA", 1);
+    pub const HANGUL_SYLLABLE_TYPE_TRAILING_JAMO_V1: ResourceKey =
+        resource_key!(uniset, "hst=T", 1);
+    pub const HANGUL_SYLLABLE_TYPE_VOWEL_JAMO_V1: ResourceKey = resource_key!(uniset, "hst=V", 1);
+    pub const INDIC_POSITIONAL_CATEGORY_BOTTOM_V1: ResourceKey =
+        resource_key!(uniset, "InPC=Bottom", 1);
+    pub const INDIC_POSITIONAL_CATEGORY_BOTTOM_AND_LEFT_V1: ResourceKey =
+        resource_key!(uniset, "InPC=BottomLeft", 1);
+    pub const INDIC_POSITIONAL_CATEGORY_BOTTOM_AND_RIGHT_V1: ResourceKey =
+        resource_key!(uniset, "InPC=BottomRight", 1);
+    pub const INDIC_POSITIONAL_CATEGORY_LEFT_V1: ResourceKey =
+        resource_key!(uniset, "InPC=Left", 1);
+    pub const INDIC_POSITIONAL_CATEGORY_LEFT_AND_RIGHT_V1: ResourceKey =
+        resource_key!(uniset, "InPC=LeftRight", 1);
+    pub const INDIC_POSITIONAL_CATEGORY_NA_V1: ResourceKey = resource_key!(uniset, "InPC=NA", 1);
+    pub const INDIC_POSITIONAL_CATEGORY_OVERSTRUCK_V1: ResourceKey =
+        resource_key!(uniset, "InPC=Overstruck", 1);
+    pub const INDIC_POSITIONAL_CATEGORY_RIGHT_V1: ResourceKey =
+        resource_key!(uniset, "InPC=Right", 1);
+    pub const INDIC_POSITIONAL_CATEGORY_TOP_V1: ResourceKey = resource_key!(uniset, "InPC=Top", 1);
+    pub const INDIC_POSITIONAL_CATEGORY_TOP_AND_BOTTOM_V1: ResourceKey =
+        resource_key!(uniset, "InPC=TopBottom", 1);
+    pub const INDIC_POSITIONAL_CATEGORY_TOP_AND_BOTTOM_AND_LEFT_V1: ResourceKey =
+        resource_key!(uniset, "InPC=TopBotLeft", 1);
```

Comment:
This looks like one of the keys you shortened to fit in 16 characters.  Can you add a test to assert that PpucdDataProvider returns valid data when you ask for a key like this?

Reply:
I added the test, but it reveals that the manual truncation of the subcategory string to fit in 16 characters actually causes an error. The `ResourceKey`'s subcategory string is used to select the correct `UnicodeProperty` from the PPUCD provider via `UnicodeProperty`'s `name` field. The `name` field is computed automatically by the parser as `<propname>=<propval>`.

The 16 character restriction is a problem in uniquely defining the specific UnicodeProperty (aka UnicodeSet) for a specific enum property name=value combination.

I guess the options are
1. revisit the TinyStr16 type for subcategory that causes 16 char limit
2. create a map that stores the mapping of  the new manual truncation string to the original `<propname>=<propval>` string, and use that map when doing a lookup of the correct `UnicodeProperty` from the data provider when given a ResourceKey

Option 2 looks preferable.

Reply:
In particular, we can make option 2 efficient by only storing maps for 16+ string lengths and use them only for keys longer than 16 :)

Reply:
I'm sorry, I really don't understand what the problem is.  Is there an *example* of a `<propname>=<propval>` that is *not* unique when truncated to 16 characters?

Option 3 would be to create a very simple hashing function that maps `<propname>=<propval>` that is 16 characters long.

It's essential that ResourceKeys are const-constructible and easy to scan for in assembly code, so revisiting that type is not an option I want to explore until all other options have been exhausted.

Reply:
To be clear: when I say "truncate" I mean literally slicing off all characters after the 16th character.  "foo=abcdefghijklmnop" truncates to "foo=abcdefghijkl".

Reply:
There's `InPC=Top_And_Bottom_And_Left` & `InPC=Top_And_Bottom_And_Right`; `jg=Manichaean_Sadhe` & `jg=Manichaean_Samekh`; `jg=Manichaean_Ten` & `jg=Manichaean_Teth`. I'm not sure if there are any other instances where truncating is ambiguous.

Reply:
Okay. I'm convinced. I tried a few other hashing functions and couldn't reliably eliminate the collisions.  Just make a map (perhaps using Manish's new VecMap) that lives inside provider_ppucd that maps from strings that exceed 16 characters to the fixed strings in the uniset::props provider.

Reply:
@Manishearth suggested the PHF crate as another alternative.

Reply:
@zbraniecki and @Manishearth also suggested using a codegen'd enum.